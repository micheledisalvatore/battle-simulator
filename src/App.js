import React, { useReducer, useEffect } from 'react';

import Dice from './components/dice'
import Bar from './components/bar'
import Player from './components/player'
import Monster from './components/monster'

import { H1, GameField, PlayerSide, MonsterSide, Dices, BattleField, Score, Attack, GameOver } from './App.styles'
import { initialState, battleReducer } from './store/battle-store';

const App = () => {
  const [state, dispatch] = useReducer(battleReducer, initialState)

  const handleAttach = () => {
    if (!state.attackInProgress) {
      dispatch({ type: 'startAttack' });
    }
  }

  useEffect(() => {
    if (state.attackInProgress) {
      window.setTimeout(() => {
        dispatch({ type: 'completeAttack' });
      }, 1000)
    }
  }, [state.attackInProgress])

  return (
    <>
      <header>
        <H1>Battle simulator</H1>
      </header>
      <GameField>
        <PlayerSide>
          <Player lost={state.player.score < 1} />
          <Bar level={state.player.score} last={state.player.lastAttack.score} />
          <Dices>
            <Dice final={state.player.lastAttack.dice1} isInProgress={state.attackInProgress} />
            <Dice final={state.player.lastAttack.dice2} isInProgress={state.attackInProgress} />
          </Dices>
        </PlayerSide>
        <MonsterSide>
          <Dices>
            <Dice final={state.monster.lastAttack.dice1} isInProgress={state.attackInProgress} />
            <Dice final={state.monster.lastAttack.dice2} isInProgress={state.attackInProgress} />
          </Dices>
          <Bar level={state.monster.score} last={state.monster.lastAttack.score} />
          <Monster lost={state.monster.score < 1} />
        </MonsterSide>
      </GameField>
      <BattleField>
        <Score>
          { state.status === 'in-progress' && (
            <>
              {state.player.lastAttack.winner && state.monster.lastAttack.score > 0 && `You hit for ${state.monster.lastAttack.score}!`}
              {state.monster.lastAttack.winner && state.player.lastAttack.score > 0 && `Monster hits for ${state.player.lastAttack.score}!`}
              {state.monster.lastAttack.score === state.player.lastAttack.score && 'Tie!'}
            </>
          )}
          { state.status === 'over' && (
            <GameOver winner={state.player.score > 0}>{state.player.score > 0 ? 'You win' : 'Game over'}</GameOver>
          )}
        </Score>
        <Attack type="button" onClick={handleAttach} data-testid="attack-button">Attack</Attack>
      </BattleField>
    </>
  );
}

export default App;
