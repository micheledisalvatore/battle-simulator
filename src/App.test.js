import React from 'react';
import { render, screen, fireEvent, waitForDomChange } from '@testing-library/react';
import { mockRandom, resetMockRandom } from 'jest-mock-random'
import App from './App';

jest.mock('./components/dice', () => jest.fn(({ final, isInProgress }) => (
  <div
    data-testid="dice"
    data-final={final}
    data-is-in-progress={isInProgress}
  />
)))

jest.mock('./components/bar', () => jest.fn(({ level, last }) => (
  <div
    data-testid="bar"
    data-last={last}
    data-level={level}
  />
)))

jest.mock('./components/player', () => jest.fn(({ lost }) => (
  <div
    data-testid="player"
    data-lost={lost}
  />
)))

jest.mock('./components/monster', () => jest.fn(({ lost }) => (
  <div
    data-testid="monster"
    data-lost={lost}
  />
)))

jest.useFakeTimers()

beforeAll(() => {
  mockRandom([0.12, 0.13, 0.14, 0.15])
})

afterAll(() => {
  resetMockRandom()
})

test('renders the component after 2 attacks', () => {
  const { container } = render(<App />)
  waitForDomChange()
  fireEvent.click(screen.getByTestId('attack-button'))
  jest.runAllTimers()

  expect(screen.getAllByTestId('bar')[0]).toHaveAttribute('data-last', "4")
  expect(screen.getAllByTestId('bar')[0]).toHaveAttribute('data-level', "96")
  expect(screen.getAllByTestId('dice')[0]).toHaveAttribute('data-final', "1")
  expect(screen.getAllByTestId('dice')[1]).toHaveAttribute('data-final', "2")

  expect(screen.getAllByTestId('bar')[1]).toHaveAttribute('data-last', "0")
  expect(screen.getAllByTestId('bar')[1]).toHaveAttribute('data-level', "100")
  expect(screen.getAllByTestId('dice')[2]).toHaveAttribute('data-final', "3")
  expect(screen.getAllByTestId('dice')[3]).toHaveAttribute('data-final', "4")

  fireEvent.click(screen.getByTestId('attack-button'))
  jest.runAllTimers()
  waitForDomChange()
  expect(screen.getAllByTestId('bar')[0]).toHaveAttribute('data-last', "4")
  expect(screen.getAllByTestId('bar')[0]).toHaveAttribute('data-level', "92")
  expect(screen.getAllByTestId('bar')[1]).toHaveAttribute('data-last', "0")
  expect(screen.getAllByTestId('bar')[1]).toHaveAttribute('data-level', "100")

  expect(container).toMatchSnapshot()
})
