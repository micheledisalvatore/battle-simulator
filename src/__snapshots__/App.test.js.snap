// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component after 2 attacks 1`] = `
<div>
  <header>
    <h1
      class="sc-AxjAm jWlVxK"
    >
      Battle simulator
    </h1>
  </header>
  <div
    class="sc-AxirZ bVCxIE"
  >
    <div
      class="sc-AxiKw cMaQPn"
    >
      <div
        data-lost="false"
        data-testid="player"
      />
      <div
        data-last="4"
        data-level="92"
        data-testid="bar"
      />
      <div
        class="sc-AxhUy bVNnww"
      >
        <div
          data-final="1"
          data-is-in-progress="false"
          data-testid="dice"
        />
        <div
          data-final="2"
          data-is-in-progress="false"
          data-testid="dice"
        />
      </div>
    </div>
    <div
      class="sc-AxiKw sc-AxhCb ibPXJs"
    >
      <div
        class="sc-AxhUy bVNnww"
      >
        <div
          data-final="3"
          data-is-in-progress="false"
          data-testid="dice"
        />
        <div
          data-final="4"
          data-is-in-progress="false"
          data-testid="dice"
        />
      </div>
      <div
        data-last="0"
        data-level="100"
        data-testid="bar"
      />
      <div
        data-lost="false"
        data-testid="monster"
      />
    </div>
  </div>
  <div
    class="sc-AxgMl cbTHOa"
  >
    <h2
      class="sc-AxheI hIhWkz"
    >
      Monster hits for 4!
    </h2>
    <button
      class="sc-Axmtr igOteH"
      data-testid="attack-button"
      type="button"
    >
      Attack
    </button>
  </div>
</div>
`;
